HeatEquationSolver/heat_equation_cn.py

```diff
@@ -0,0 +1,127 @@
+import numpy as np
+from matplotlib import pyplot as plt
+from matplotlib.animation import FuncAnimation
+
+# setting up parameters
+length = 1
+nodes = 100
+time = 30
+k = 0.01
+initial_temp = 25
+
+
+def left_temp_func(temp, t):
+    """
+    Generates the temperature at the left boundary of the rod
+
+    :param temp: initial temperature of the rod at t=0
+    :param t: array consisting of time steps from t=0 to t= total time
+    :return: the temperature at the left boundary
+    """
+    return 20 * np.sin(np.pi * t) + temp
+
+
+def right_temp_func(temp, t):
+    """
+    Generates the temperature at the right boundary of the rod
+
+    :param temp: initial temperature of the rod at t=0
+    :param t: array consisting of time steps from t=0 to t= total time
+    :return: the temperature at the right boundary
+    """
+    return t + temp
+
+
+def build_tridiagonal_matrix(a, b, c, nodes):
+
+    matrix = np.zeros((nodes, nodes))
+    np.fill_diagonal(matrix, b)
+    np.fill_diagonal(matrix[1:], a)
+    np.fill_diagonal(matrix[:, 1:], c)
+
+    return matrix
+
+
+def solve_heat_equation_cn(length, nodes, time, k, initial_temp):
+    """
+    Solves the heat equation using the explicit method
+
+    :param length: length of the rod
+    :param nodes: number of spatial nodes across the rod
+    :param time: the total time
+    :param k: heat diffusivity constant
+    :param initial_temp: initial temperature of the rod at t=0
+    :return: the solved 2d matrix consisting the temperature distribution over time
+    """
+    dx = length / (nodes-1)
+    dt_max = 0.5 * (dx**2) / k  # calculating dt using a formula (to ensure stability)
+    dt = 0.8 * dt_max
+    time_step = int(time/dt)
+    t = np.linspace(0, time, time_step)
+
+    alpha = k * dt / (2 * dx**2)
+    a = -alpha
+    b = 1 + 2 * alpha
+    c = -alpha
+
+    u = np.zeros((time_step, nodes))
+
+    u[0, :] = initial_temp
+    u[:, 0] = left_temp_func(initial_temp, t)
+    u[:, -1] = right_temp_func(initial_temp, t)
+
+    lhs = build_tridiagonal_matrix(a, b, c, nodes-2)
+    rhs = np.zeros(nodes-2)
+
+    for tau in range(0, time_step - 1):
+        rhs[0] = alpha * (u[tau, 0] + u[tau+1, 0]) + (1 - 2 * alpha) * u[tau, 1] + alpha * u[tau, 2]
+
+        for i in range(1, nodes-3):
+            rhs[i] = alpha * u[tau, i] + (1 - 2 * alpha) * u[tau, i+1] + alpha * u[tau, i+2]
+
+        rhs[-1] = alpha * (u[tau, -1] + u[tau+1, -1]) + (1 - 2 * alpha) * u[tau, -2] + alpha * u[tau, -3]
```

Comment:
i feel like there's something wrong with this

Reply:
You should check whether the explicit method and the CN method give you similar solutions.

Reply:
it does with slight differences !